### DISCUSSION TITLE
Naming a template-defined asset transform

### DISCUSSION THREAD

**Discussion by evanwarner** [craftcms/cms#8428] [Ideas]
It would be great if asset transforms defined in template code could include a name that would be used for the transform directory. Getting folders named things like `_800xAUTO_crop_center-center` isn't very desirable.

**Comment by brandonkelly** [comment]
Like this idea, @andris-sevcenko. `name` should be another option alongside `width`, `height`, `mode`, etc., and when generating transform folders, just use whatever `AssetTransform::$name` is set to if it's set (rather than worrying about whether it has an ID (assuming that's what it's doing)), otherwise come up with the auto-generated name.

**Comment by andris-sevcenko** [comment]
The logic behind those folder names is so that Craft knows what's the transform format. This helps Craft re-use created transforms if the format matches. Granted, this would be a rare circumstance, so not much of a loss there.

I also like the idea, but it's not as simple as it seems at face value. Honestly, the current approach is inherently resistant against conflicts, while the proposed approach is anything but.

1) What to do if a name parameter is used with a value that matches a named transform? The probable answer here is "just use the generated name in that case", but that just postpones the problem - when rendering the same image next time, how would Craft know if it has to look up using the named transform folder name or the generated folder name?

2) If there is a named transform in the template and a named transform is created later with the same name - how would potential conflicts be resolved then for transforms that have already been created?

3) What if the same name is used for two different parameters in two different templates? Which is more important and what to do with the other?

Note, that most of the solutions to the problems above will probably have a version of "just use the generic descriptive folder name, then" which is still not solving the problem of ditching those folder names.

We'll definitely discuss this, but this is not a trivial problem to solve, imo.

**Comment by brandonkelly** [comment]
_Conflicts Schmonflicts._

**Comment by angrybrad** [comment]
_You think this bad conflict?_

**Comment by andris-sevcenko** [comment]
❤️

**Comment by narration-sd** [comment]
Andris, I think maybe you should help Brad with his 'wegetables'. 

Or maybe Brandon :)

(spelling from an old family joke)

**Comment by evanwarner** [comment]
Andris, here are some thoughts on your points. I don't know all the ins-and-outs of how transforms are calculated and stored, so excuse any flawed logic here.

First, I think all three are slightly "RTFM" moments for developers. All cases are kind of similar to redefining a variable. If I do that, I can expect my code not to work the way I expect.

1. What if a Twig-defined transform were treated as an override to a CP-defined one by the same name? It could be a cumulative override, only overriding the options declared in the code, or a full override completely replacing the CP definition. I actually think this could be super useful in some cases, allowing developers to conditionally apply different versions of a transform to an image if they detect a certain thing about the image.

2. This is the reverse of situation 1, so the above still applies. Here though, it's the situation of someone in the CP creating a transform they didn't realize was in the code. My immediate reaction to that is "well, anyone changing these things should have knowledge of the full picture." I'm not sure how much your product features should try to account for poor organization and communication of users. It's definitely an avoidable problem. But also, if the Twig-defined transform is a full override, theoretically creating a duplicate in the CP shouldn't actually break any existing code.

3. Would these two ever collide? If the definitions are in different templates, only one should exist on any page load, correct?

**Comment by andris-sevcenko** [comment]
Evan, your solutions are really good ideas, but I'm guilty of assuming everyone knows how that worked behind the scene, so here's a little in-depth description of the process.

When you request a transform for an Asset in the template - first thing that Craft does is determine the location folder for that transform. For named transforms the folder name is the handle of the transform, for unnamed transforms that folder name is generated by using the transform parameters.

Then, Craft looks up if an entry exists for that Asset in that folder using the `craft_assettransformindex` table. This helps Craft avoid unnecessary requests to figure out if a transform should be created or not.

So this is where we run into the issue - once a transform is generated, the entry is stored in that same DB table. This means that if you call a transform for an Asset that has already been generated, even if you override some of the parameters in the template, it's not going to matter, because a transform has been generated for that folder and Craft is just going to serve up that.

-----

That being said, your use case for overriding transforms in case of different things is a really awesome example on why we should probably re-vamp the whole process.

**Comment by evanwarner** [comment]
Yep, that makes sense that the existing record would just prevent the Twig-defined transform from even being considered. Spitballing some other thoughts that may trigger ideas for you:

I assume when you make changes to a CP-defined transform, that triggers some sort of invalidation that wipes out all the records and transformed images tied to that transform. With Twig-defined transforms, changing the transform options naturally changes the folder name and acts as its own invalidation (then maybe a Craft cleanup task purges the old folder at some point?).

Might the solution come in two parts:
1. Twig-defined transforms get records in the `assettransforms` table (created/updated anytime you call `getUrl(myTwigTransform)` etc.). They are allowed to have the same `name`/`handle` as CP-defined transforms. Twig-defined transforms, however, get a unique key similar to how the `{% cache %}` tag can generate its own or use a user-defined one. This would allow a developer to define as many different versions of a same-named transform as they want; the keys distinguish them all. Essentially this just gives some permanence to Twig-defined transforms, and they might just happen to store files in the same folder as another transform.
2. When Craft is looking to see if a transform exists in the `assetstransformindex` table, it check the record's `dateUpdated` against the provided `assettransform`'s `dateUpdated` and if the latter is newer it re-transforms the image. (This is to help invalidate images created from Twig-defined transforms since there's no CP activity to trigger it.)
